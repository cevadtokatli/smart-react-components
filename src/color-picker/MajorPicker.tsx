import React from "react"
import DOMHelper from "../helper/DOMHelper"
import ColorPickerHelper from "../helper/ColorPickerHelper"
import size from "../constants/color-picker/size"
import MajorPickerElement from "./MajorPickerElement"
import MajorDraggerElement from "./MajorDraggerElement"

export interface MajorPickerOnRef {
    setValue: (x:number, y:number) => void 
}

interface Props {
    breakpoint: string
    rgba: React.RefObject<number[]>
    processStart: () => void
    processEnd: () => void
    majorPickerProcess: (value:string, rgb:number[]) => void
    majorPickerBgEl: React.RefObject<HTMLDivElement>
    majorPickerDraggerEl: React.RefObject<HTMLSpanElement>
    onRef: (e:MajorPickerOnRef) => void
}

const MajorPicker: React.FC<Props> = ({breakpoint,rgba,processStart,processEnd,majorPickerProcess,majorPickerBgEl,majorPickerDraggerEl,onRef}) => {
    const pickerEl = React.useRef<HTMLDivElement>(null)

    /**
     * Adds mouse and touch events.
     * Calls the process methods.
     * 
     * @param e 
     */
    const start = (e:React.MouseEvent|React.TouchEvent): void => {
        DOMHelper.addEventListener(document, ["mousemove", "touchmove"], move)
        DOMHelper.addEventListener(document, ["mouseup", "touchend"], end)
        processStart()
        process(e as any as MouseEvent)
    }
    
    /**
     * Calls the process method.
     *
     * @param e 
     */
    const move = (e:Event) => process(e as MouseEvent)
    
    /**
     * Removes mouse and touch events.
     * Calls the end process method.
     */
    const end = (): void => {
        DOMHelper.removeEventListener(document, ["mousemove", "touchmove"], move)
        DOMHelper.removeEventListener(document, ["mouseup", "touchend"], end)
        processEnd()
    }

    /**
     * Sets position of the major picker dragger.
     * Calls the set value method.
     * 
     * @param e 
     */
    const process = (e:MouseEvent): void => {
        const {left, top, x, y} = ColorPickerHelper.getDraggerNewPosition(e, pickerEl.current, size[breakpoint].majorDraggerSize)
        majorPickerDraggerEl.current.style.left = `${left}px`
        majorPickerDraggerEl.current.style.top = `${top}px`
        setValue(x, y)
    }

    /**
     * Sets value by the given coordinators.
     * 
     * @param x 
     * @param y 
     */
    const setValue = (x:number, y:number): void => {
        const rgb = [rgba.current[0], rgba.current[1], rgba.current[2]]
        const width = size[breakpoint].majorPickerWidth
        const height = size[breakpoint].majorPickerHeight

        for(let i in rgb) {
            const v = rgb[i]
            let netVal
            if(v == 255)
                netVal = Math.abs(Math.round(((255 / height) * y) - 255))
            else {
                const topNetVal = Math.round((height - y) * (v / height))
                const leftVal = Math.round((width - x) * ((255 - v) / width))
                const leftNetVal = Math.abs(Math.round((height - y) * (leftVal / width)))
                netVal = topNetVal + leftNetVal
            }
            rgb[i] = netVal
        }

        majorPickerProcess(`rgba(${rgb[0]}, ${rgb[1]}, ${rgb[2]}, ${rgba.current[3]})`, rgb)
    }

    React.useEffect(() => {
        onRef({setValue}) 
    }, [])

    return (
        <MajorPickerElement ref={pickerEl} onMouseDown={start} onTouchStart={start}>
            <div ref={majorPickerBgEl}>
                <div className="src-color-picker-gradient-light">
                    <div className="src-color-picker-gradient-dark src-color-picker-picker-shadow">
                        <MajorDraggerElement ref={majorPickerDraggerEl} />
                    </div>
                </div>
            </div>
        </MajorPickerElement>
    )
}

export default MajorPicker