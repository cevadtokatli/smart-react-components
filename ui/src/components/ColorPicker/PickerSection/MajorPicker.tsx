import { getColor } from '@smart-react-components/core/util/color'
import React from 'react'
import ColorPickerProps from '../../../context/ColorPickerProps'
import { PickerProps } from '../../../types/color-picker'
import { calculateDraggerPosition, isLight } from '../../../util/color-picker'
import MajorDraggerElement from './MajorDraggerElement'
import MajorPickerElement from './MajorPickerElement'

const MajorPicker: React.FC<PickerProps> = ({ color, onDrag, onEnd, onRef, onStart }) => {
  const props = React.useContext(ColorPickerProps)
  const pickerEl = React.useRef<HTMLDivElement>(null)
  const bgEl = React.useRef<HTMLDivElement>(null)
  const draggerEl = React.useRef<HTMLSpanElement>(null)

  const handleChange = () => {
    if (!pickerEl.current) {
      return
    }

    let background: string
    let isDark: boolean
    let left: number
    let top: number

    const draggerHalfSize = draggerEl.current.offsetWidth / 2

    if (!color.current) {
      background = '#F0O'
      isDark = true
      left = draggerHalfSize * -1
      top = draggerHalfSize * -1
    } else {
      const rgb = color.current.rgb().color
      const hsl = color.current.hsl().color
      const [r, g, b] = rgb
      const [h] = hsl
      const maxColor = Math.max(r, g, b)
      const minColor = Math.min(r, g, b)

      const pickerSize = pickerEl.current.offsetWidth
      const topNetVal = Math.abs(Math.round(((pickerSize / 255) * maxColor) - pickerSize))
      const leftVal = Math.abs(Math.round(((pickerSize / 255) * minColor) - pickerSize))
      const leftNetVal = leftVal - Math.abs(Math.round((topNetVal / maxColor) * minColor))

      background = `hsl(${h}, 100%, 50%)`
      isDark = isLight(color.current, false)
      left = leftNetVal - draggerHalfSize
      top = topNetVal - draggerHalfSize
    }

    draggerEl.current.setAttribute('data-src-is-dark', String(isDark))
    draggerEl.current.setAttribute('style', `left: ${left}px; top: ${top}px;`)
    pickerEl.current.style.background = background
  }

  const handleDrag = (e?: MouseEvent) => {
    let x: number
    let y: number

    if (e) {
      let left: number
      let top: number

      ({ left, top, x, y } = calculateDraggerPosition(e, pickerEl.current, draggerEl.current.offsetWidth))

      draggerEl.current.style.left = `${left}px`
      draggerEl.current.style.top = `${top}px`
    } else {
      const draggerHalfSize = draggerEl.current.offsetWidth / 2
      x = draggerEl.current.offsetLeft - draggerHalfSize
      y = draggerEl.current.offsetTop - draggerHalfSize
    }

    const currentColor = color.current ?? getColor('#FFF')
    const [h] = currentColor.hsl().color
    const rgb = (getColor(`hsl(${h}, 100%, 50%)`).rgb() as any).color
    const pickerSize = pickerEl.current.offsetWidth

    for (const i in rgb) {
      const v = rgb[i]
      let netValue: number

      if (v === 255) {
        netValue = Math.abs(Math.round(((255 / pickerSize) * y) - 255))
      } else {
        const topNetValue = Math.round((pickerSize - y) * (v / pickerSize))
        const leftValue = Math.round((pickerSize - x) * ((255 - v) / pickerSize))
        const leftNetValue = Math.abs(Math.round((pickerSize - y) * (leftValue / pickerSize)))
        netValue = topNetValue + leftNetValue
      }

      rgb[i] = netValue
    }

    color.current = getColor(`rgba(${rgb[0]}, ${rgb[1]}, ${rgb[2]}, ${currentColor.valpha})`)

    draggerEl.current.setAttribute('data-src-is-dark', String(isLight(color.current, false)))

    onDrag()
  }

  const handleExternalDrag = () => {
    const [h] = color.current.hsl().color
    bgEl.current.style.background = `hsl(${h}, 100%, 50%)`

    handleDrag()
  }

  const handleEnd = () => {
    onEnd(handleMove, handleEnd)
  }

  const handleMove = (e: MouseEvent) => {
    handleDrag(e)
  }

  const handleStart = (e: React.MouseEvent | React.TouchEvent) => {
    onStart(handleMove, handleEnd)
    handleDrag(e as any as MouseEvent)
  }

  React.useEffect(() => {
    onRef({
      onChange: handleChange,
      onExternalDrag: handleExternalDrag,
    })
  }, [])

  return (
    <MajorPickerElement
      {...props}
      bgElRef={bgEl}
      onMouseDown={handleStart}
      onTouchStart={handleStart}
      ref={pickerEl}
    >
      <MajorDraggerElement
        {...props}
        ref={draggerEl}
      />
    </MajorPickerElement>
  )
}

export default MajorPicker
