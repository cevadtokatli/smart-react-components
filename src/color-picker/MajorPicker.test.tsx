import React from "react"
import renderTheme from "../test/renderTheme"
import MajorPicker from "./MajorPicker"

describe("<MajorPicker />", () => {
    let props

    beforeEach(() => {
        props = {
            breakpoint: "large",
            rgba: {
                current: ["0","0","0","1"]
            },
            processStart: jest.fn(),
            processEnd: jest.fn(),
            majorPickerProcess: jest.fn(),
            majorPickerBgEl: {},
            majorPickerDraggerEl: {},
            onRef: jest.fn()
        }
    })

    it("should match snapshot", () => {
        const wrapper = renderTheme(<MajorPicker {...props} />)
        expect(wrapper).toMatchSnapshot()
    })

})