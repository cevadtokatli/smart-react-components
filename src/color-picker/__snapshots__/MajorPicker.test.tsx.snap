// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MajorPicker /> should match snapshot 1`] = `
<div
  class="sc-AxjAm crxhoX"
>
  <div>
    <div
      class="src-color-picker-gradient-light"
    >
      <div
        class="src-color-picker-gradient-dark src-color-picker-picker-shadow"
      >
        <span
          class="sc-AxirZ bhwKok"
        />
      </div>
    </div>
  </div>
</div>
`;
